### PR TITLE
feat(conditions): Move the condition assignments interface to its own class

### PR DESCRIPTION
**Pull request by petervdmeer**
**Feature**
This PR makes a first step in addressing the feature described in issue #10213

## Summary
This PR makes a first step in splitting the ConditionSet into expressions and assignments; it splits the interface, and makes sure that various users of conditions call the correct version of the class (depending on if they want to query, or if they want to apply conditions).

This PR is mostly groundwork to allow introducing a parser that generates an Abstract Syntax Tree in a follow-up PR. (I plan to upload a draft of such a follow-up PR soon after this PR gets merged.)

Splitting up the interfaces also makes it easier to write more unit tests specifically for assignments, and tests specifically for testable expressions. (The arithmetic parsing code seems to be a bit low on test-coverage.) This is also for a follow-up PR.

## Screenshots
N/A

## Usage examples
For coders:
- Use ConditionSet for conditions you want to check.
- Use ConditionAssignments if you want to modify conditions.

For content-creators:
- The Wiki already has the correct descriptions of how things are supposed to work, so just use conditions as described on the wiki.

For end-users:
- End users should not notice anything from this PR.

## Testing Done
The unit-tests have been updated, and I verified that most calls map 1:1 from the old situation to the new situation.

## Save File
N/A. This can be tested with any savefile. (But the best tests are probably the unit-tests.)

## Wiki Update
The Wiki already has separate sections for `Testable condition sets` and `Applied condition sets`, so that doesn't seem to need an update.

## Performance Impact
I expect no performance impact.

### CODE REVIEW

File: source/ConditionSet.h [hunk 1]
```diff
@@ -45,9 +45,22 @@ class ConditionSet {
...
+	// Check if the given condition values satisfy this set of expressions. First applies
+	// all assignment expressions to create any temporary conditions, then evaluates.
+	bool Test(const ConditionsStore &conditions) const;
+
+	// Get the names of the conditions that are modified by this ConditionSet.
+	std::set<std::string> RelevantConditions() const;
+
+
+
+private:
```
Comment (RisingLeaf) [inline source/ConditionSet.h:63]: please move the functions in the source file too

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
Might be a good idea to edit the comment on top of the ConditionSet class

**Comment by petervdmeer** [comment]
> Might be a good idea to edit the comment on top of the ConditionSet class

Done, I also described a bit of the next change that comes after this PR.

**Review by RisingLeaf — APPROVED** [review]
I did not look at the test cause I have absolutely no idea about the tests

File: source/ConditionAssignments.cpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+
+bool ConditionAssignments::Add(const string &name, const string &op, const string &value)
+{
+	return setToEvaluate.Add(name, op, value);
+}
+
+
+
+bool ConditionAssignments::Add(const vector<string> &lhs, const string &op, const vector<std::string> &rhs)
```
Comment (TomGoodIdea) [inline source/ConditionAssignments.cpp:92]:
```suggestion
bool ConditionAssignments::Add(const vector<string> &lhs, const string &op, const vector<string> &rhs)
```

File: source/ConditionSet.cpp [hunk 3]
```diff
@@ -280,6 +288,38 @@ bool ConditionSet::IsEmpty() const
...
+		TestApply(conditions, created);
+	return TestSet(conditions, created);
+}
+
+
+
+// Get the names of the conditions that are relevant for this ConditionSet.
+set<string> ConditionSet::RelevantConditions() const
+{
+	set<std::string> result;
```
Comment (TomGoodIdea) [inline source/ConditionSet.cpp:309]:
```suggestion
	set<string> result;
```

**Review by TomGoodIdea — COMMENTED** [review]
I'm not really sure if this split into two classes is needed for the features you mentioned. IMO it adds some complexity, and could be avoided, especially if most of the new functions are just simple operations on the underlying ConditionSet.

**Comment by petervdmeer** [comment]
> I'm not really sure if this split into two classes is needed for the features you mentioned.

It is needed for an AST parser, because parsing for `Applied condition sets` and `Testable condition sets` (as the Wiki calls them) is different.

Having an AST parser has certain benefits, for example:
- syntax-errors can be detected at load-time. (The current ConditionSet code detects certain errors at runtime.)
- it could decide at load-time that `8 + 3 * 5` evaluates to `23`, and just immediately store `23` instead of the expression/calculation (saving calculations later at runtime).
- `someVar and never` can be optimized to `never` at load-time, removing unnecessary lookups for `someVar` at runtime.
- that we lose the overhead of the intermediate `const ConditionsStore &created` conditions when only testing/evaluating conditions.
- condition-monitoring can be implemented a bit more cleanly.
- we can return 'int64_t' results from ConditionSet, instead of only boolean. Allowing conditions to be used for more than just yes/no questions.

> IMO it adds some complexity

I feel that large part of this complexity comes from the current situation of having both `Applied condition sets` and `Testable condition sets` combined into a single class. This complexity should be gone when all support for assignment operators are moved out of ConditionSet into ConditionAssignments.

> and could be avoided, especially if most of the new functions are just simple operations on the underlying ConditionSet.

Agreed, I plan to avoid this by introducing an ASR parser (in a follow-up PR); the ASR parser code will move all assignment operators out of the ConditionSet class, into the ConditionExpression class, removing the need to forward assignment-related operations to the ConditionSet.

File: source/ConditionSet.cpp [hunk 4]
```diff
@@ -272,6 +272,14 @@ void ConditionSet::Save(DataWriter &out) const
 
 
 
+void ConditionSet::MakeNever()
+{
+	// Add the equivalent "never" condition, `"'" != 0`.
+	Add("has", "'");
```
Comment (tibetiroka) [inline source/ConditionSet.cpp:278]: Is there anything preventing a plugin from declaring the `'` condition as true?

File: tests/unit/src/test_conditionSet.cpp [hunk 5]
```diff
@@ -76,22 +76,22 @@ SCENARIO( "Extending a ConditionSet", "[ConditionSet][Creation]" ) {
 		REQUIRE( set.IsEmpty() );
 
 		THEN( "no expressions are added from empty nodes" ) {
-			set.Add(DataNode{});
+			set.Load(DataNode{});
 			REQUIRE( set.IsEmpty() );
-			AND_THEN( "a log message is printed to assist the user" ) {
-				REQUIRE( warnings.Flush() == validationWarning );
-			}
+			// AND_THEN( "a log message is printed to assist the user" ) {
```
Comment (tibetiroka) [inline tests/unit/src/test_conditionSet.cpp:81]: I'm not super happy about these disabled tests.

File: source/ConditionSet.cpp (hunk 4 above)
Comment (petervdmeer) [reply source/ConditionSet.cpp:278]: No, I don't think anything is preventing a plugin from declaring the '\'' condition to be true. This is not a new issue, this issue only moved from Mission.cpp towards here.
I already planned to solve this in a follow-up PR, and I added 2 TODO items as reminders so that it doesn't get forgotten.

File: tests/unit/src/test_conditionSet.cpp (hunk 5 above)
Comment (petervdmeer) [reply tests/unit/src/test_conditionSet.cpp:81]: I agree, we should have tests for "empty" and "invalid" conditions, both for ConditionSets and for ConditionAssignments. I was able to restore the test for "invalid" conditions after dealing with some string beautification functionality. Will check later what I can do for the other tests.

File: source/ConditionSet.cpp (hunk 4 above)
Comment (tibetiroka) [reply source/ConditionSet.cpp:280]: I've been thinking. According to our data format, we can't parse
> " `

as a token. It can't be quoted in quotes nor backticks, and it has a whitespace so it must be quoted.

It's a bit hacky, but this is something plugins can't possibly override. Well, unless there are some weird edge cases within our data parsing code.

Comment (RisingLeaf) [reply source/ConditionSet.cpp:280]: Just have a check for one specific word and don't do hacks

File: tests/unit/src/test_conditionSet.cpp (hunk 5 above)
Comment (petervdmeer) [reply tests/unit/src/test_conditionSet.cpp]: I restored both checks, but some condition-users actually try to load empty conditions (which now gives an error-message). I'll update those callers not to load the condition when the condition is empty.

Comment (petervdmeer) [reply tests/unit/src/test_conditionSet.cpp]: Also updated the substitutions in conversations and a branch-statement in the data-files that was missing conditions. I believe this is solved now.

File: source/ConditionSet.cpp (hunk 4 above)
Comment (petervdmeer) [reply source/ConditionSet.cpp:282]: > Just have a check for one specific word and don't do hacks

Or we just validate if condition-names are valid by using a regular expression when loading conditions (and their names).

Comment (petervdmeer) [reply source/ConditionSet.cpp:282]: Probably not in this PR though, but the TODO item is at least created.

File: source/ConditionSet.cpp [hunk 6]
```diff
@@ -247,6 +247,8 @@ ConditionSet::ConditionSet(const DataNode &node)
 void ConditionSet::Load(const DataNode &node)
 {
 	isOr = (node.Token(0) == "or");
+	if(! node.HasChildren())
```
Comment (tibetiroka) [inline source/ConditionSet.cpp:250]:
```suggestion
	if(!node.HasChildren())
```

**Comment by tibetiroka** [comment]
That segfault is making me feel uncomfortable, especially that I can't reproduce it.

**Comment by petervdmeer** [comment]
> That segfault is making me feel uncomfortable, especially that I can't reproduce it.

Agreed, it happened on Windows twice by now. But it doesn't seem to happen on Linux builds.

**Comment by tibetiroka** [comment]
Unfortunately, I don't have access to a windows machine that can run the game. The best I have is a VM that can compile it, but doesn't support OpenGL properly so the game doesn't even launch.

File: tests/unit/src/test_conditionAssignments.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+	OutputSink warnings(std::cerr);
+
+	GIVEN( "empty ConditionAssignments" ) {
+		auto set = ConditionAssignments{};
+		REQUIRE( set.IsEmpty() );
+
+		THEN( "no assignments are added from empty nodes" ) {
+			const std::string validationWarning = "Error: Loading empty (sub)condition:\n";
+			set.Load(DataNode{});
```
Comment (warp-core) [inline tests/unit/src/test_conditionAssignments.cpp:64]: Running the unit tests locally (built with clang) this call results in a vector subscript out of bounds error.
This empty data node has no tokens. When the node is passed to `ConditionSet::Load`, it expects at least one token, so it calls `node.Tokens(0)`, which either fails a debug assertion or results in a segfault, because there isn't actually anything there.
I would suggest that this test case should just be removed: DataFile should never produce a DataNode with no tokens, such an object has no meaning. It real DataNodes can actually be empty, we have a much bigger problem.

Comment (petervdmeer) [reply tests/unit/src/test_conditionAssignments.cpp]: Thanks for finding this!

I changed it to be a node without any children, because that should also register as empty.

**Comment by petervdmeer** [comment]
> That segfault is making me feel uncomfortable, especially that I can't reproduce it.

This is now fixed, with some help from warp-core.

File: data/korath/korath missions.txt [hunk 8]
```diff
@@ -419,6 +419,7 @@ mission "Efreti: Distress Call 4"
 				to display
 					"flagship crew" <= 1
 			branch accept
+				1 == 1
```
[the remark this answers is not in the document]
Comment (tibetiroka) [reply data/korath/korath missions.txt:422]: I think the empty branch is no longer true. (Based on some discussion around a wiki pr, this wasn't a documented/expected feature to begin with, though I'm slightly worried about plugins subtly breaking.)

[the remark this answers is not in the document]
Comment (petervdmeer) [reply data/korath/korath missions.txt:422]: I wouldn't expect _this_ PR to change the behaviour of an empty branch (it should only introduce a warning).
But what would be the proper construct to use here? Should this just be accept without the branch? Or should this always have a false condition?

[the remark this answers is not in the document]
Comment (tibetiroka) [reply data/korath/korath missions.txt:422]: I think it previously defaulted to a true condition, so that's what we should continue to use.

[the remark this answers is not in the document]
Comment (petervdmeer) [reply data/korath/korath missions.txt]: I removed this change from this PR. Checking on empty conditions under branches seems to be quite a big topic that deserves it's own ticket and PR.

File: source/Conversation.cpp [hunk 9]
```diff
@@ -160,7 +160,10 @@ void Conversation::Load(const DataNode &node)
 			// Don't merge "branch" nodes with any other nodes.
 			nodes.emplace_back();
 			nodes.back().canMergeOnto = false;
-			nodes.back().conditions.Load(child);
+			// Let's support empty conditions for now as well.
+			// Not fully officially supported, but not fully unsupported either...
```
Comment (tibetiroka) [inline source/Conversation.cpp:164]:
```suggestion
			// Maintenance note: Empty conditions might have to be removed in the future,
			// so their support is unofficial.
```

File: source/ConditionAssignments.h [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONDITION_ASSIGNMENTS_H_
+#define CONDITION_ASSIGNMENTS_H_
```
Comment (tibetiroka) [inline source/ConditionAssignments.h:17]:
```suggestion
#pragma once
```

File: source/ConditionAssignments.h [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+
+private:
+	// Add complex conditions having multiple operators, including parentheses.
+	bool Add(const std::vector<std::string> &lhs, const std::string &op, const std::vector<std::string> &rhs);
+
+	ConditionSet setToEvaluate;
+};
+
+#endif
```
Comment (tibetiroka) [inline source/ConditionAssignments.h:69]:
```suggestion
```

File: source/ConditionSet.cpp [hunk 12]
```diff
@@ -272,6 +274,16 @@ void ConditionSet::Save(DataWriter &out) const
 
 
 
+void ConditionSet::MakeNever()
+{
+	// Add the equivalent "never" condition, `"'" != 0`.
+	// TODO: change the ConditionSet to contain "false" literal, instead of a comparison that should return "false".
+	// TODO: Validate if condition-names are valid. (The "'" character by itself should not pass as a valid condition.)
+	Add("has", "'");
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/ConditionSet.cpp:282]: yes; https://github.com/endless-sky/endless-sky/blob/f0298e3354db939b5370b0b9ada9aac31a64beab/source/ConditionSet.cpp#L369

File: source/ConditionSet.h [hunk 13]
```diff
@@ -27,10 +27,11 @@ class DataWriter;
...
 
 
-// A condition set is a collection of operations on the player's set of named
-// "conditions". This includes "test" operations that just check the values of
-// those conditions, and other operations that can be "applied" to change the
-// values.
+// A condition set is a collection of operations on the player's set of named "conditions"; "test" operations that just
+// check the values of those conditions, and in the future "evaluation" operations that can calculate an int64_t value
+// based on the conditions.
+// Operations that can be "applied" to change the values were historically also here, but are being moved to
```
[the remark this answers is not in the document]
Comment (petervdmeer) [reply source/ConditionSet.h:33]: Well, the actual move happens in a follow-up PR (ConditionAssignments is currently only fowarding the calls it has). So I was planning to delete this line and the next one in the follow-up PR.

**Comment by petervdmeer** [comment]
I removed "temporary testable assignment conditions", as discussed in the context of #10406. And I moved the toplevel Apply function to ConditionAssignments, because Apply is only used in ConditionAssignments now.

**Review by tibetiroka — APPROVED** [review]

**Comment by tibetiroka** [comment]
I am, but that's not the solution we went for. It's a bit of a pain to integrate a java tool into our build process.

**Comment by petervdmeer** [comment]
> I assume you're aware of ANTLR?

Yes, you mentioned ANTLR in https://github.com/endless-sky/endless-sky/issues/10213#issuecomment-2171429335 (the issue-ticket which was filed in preparation for this PR).
